packages/adobe-sign/src/services/agreement.service.ts

```diff
@@ -74,4 +79,56 @@ export class AgreementService {
   public async getAgreementSigningUrls(id: string): Promise<string> {
     return await this.adobeSignApi.getAgreementSigningUrls(id);
   }
+
+  public async processWebhook(webhook: Webhook): Promise<void> {
+    const adobeSignAgreementTask = this.adobeSignApi.getAgreement(webhook.agreement.id);
+    const dbAgreementTask = this.agreementsRepo.getAgreementById(webhook.agreement.id);
+
+    const adobeSignAgreement = await adobeSignAgreementTask;
+    const dbAgreement = await dbAgreementTask;
+
+    dbAgreement.webhookLogs.push(new WebhookLog({
+      event: webhook.event,
+      timestamp: this.timeService.currentTimestamp()
+    }));
+
+    if (adobeSignAgreement.status === dbAgreement.status) {
+      console.log(`AgreementService.processWebhook: Status already processed. Logging webhook and returning.`);
+      await this.agreementsRepo.put(dbAgreement);
+      return;
+    }
+
+    dbAgreement.status = adobeSignAgreement.status;
+
+    if (dbAgreement.status === AgreementStatus.Signed) {
+      console.log(`AgreementService.processWebhook: Status is complete. Fetching form data and pdf.`);
+      const templateTask = this.templatesRepo.getTemplateById(dbAgreement.adobeSignTemplateId);
+      const formDataTask = this.adobeSignApi.getAgreementFormData(adobeSignAgreement.id);
+      const pdfTask = this.adobeSignApi.getAgreementPdf(adobeSignAgreement.id);
+
+      const template = await templateTask;
+      const pdf = await pdfTask;
```

Comment:
Not sure if you'd want to do the same Promise.all() treatment here since `formData` is used further below. Maybe kick that one off first and group the other two?

```suggestion
      const formDataTask = this.adobeSignApi.getAgreementFormData(adobeSignAgreement.id);
      const [template, pdf] = await Promise.all([
        this.templatesRepo.getTemplateById(dbAgreement.adobeSignTemplateId),
        this.adobeSignApi.getAgreementPdf(adobeSignAgreement.id)
      ]);
```

Not sure how much of a micro-optimization that would be as opposed to the readability of just tossing all 3 requests in there. Up to you.

Reply:
I did this. Makes sense.